src/librbd/ImageCtx.h

```diff
@@ -69,18 +69,26 @@ namespace librbd {
 
     /**
      * Lock ordering:
-     * owner_lock, md_lock, cache_lock, snap_lock, parent_lock, refresh_lock,
-     * object_map_lock, async_op_lock
+     *
+     * owner_lock, md_lock, cache_lock, snap_lock, features_lock,
+     * parent_lock, refresh_lock, object_map_lock, async_op_lock
      */
     RWLock owner_lock; // protects exclusive lock leadership updates
     RWLock md_lock; // protects access to the mutable image metadata that
-                   // isn't guarded by other locks below
-                   // (size, features, image locks, etc)
+                   // isn't guarded by other locks below, and blocks writes
+                   // when held exclusively, so snapshots can be consistent.
+                   // Fields guarded include:
+                   // flush_encountered
+                   // total_bytes_read
+                   // exclusive_locked
+                   // lock_tag
+                   // lockers
     Mutex cache_lock; // used as client_lock for the ObjectCacher
-    RWLock snap_lock; // protects snapshot-related member variables:
+    RWLock snap_lock; // protects snapshot-related member variables, and flags
+    RWLock features_lock; // protects features
```

Comment:
Could also create a new "io_lock" RWLock that is taken for read only in aio_write/aio_discard and taken for write in snap_create.  That would be enough to stop future write requests while older writes are flushed to the OSDs before the snap is created.